### ISSUE TITLE
Barrier Function improvement

### ISSUE THREAD

**Issue by LockedFlysher** [iit-DLSLab/mpx#3]
```python
def penalty(constraint, alpha=0.1, sigma=5, transition_width=0.2):
    """
    Penalty function using tanh for smooth transition

    Parameters:
    - constraint: constraint value g
    - alpha: penalty parameter
    - sigma: switching point
    - transition_width: transition width, smaller values create steeper transitions
    """

    def safe_log(x):
        x = jnp.clip(x, 1e-10, 1e6)
        return jnp.log(x)

    # Function values for the two branches
    quadratic_barrier = alpha / 2 * (jnp.square((constraint - 2 * sigma) / sigma) - jnp.ones_like(constraint))
    log_barrier = -alpha * safe_log(constraint)
    combined_barrier = quadratic_barrier + log_barrier  # Left-side combination

    # Smooth transition weight (between 0 and 1)
    # When g << σ, weight ≈ 0 (use combined_barrier)
    # When g >> σ, weight ≈ 1 (use log_barrier)
    weight = 0.5 * (1 + jnp.tanh((constraint - sigma) / transition_width))

    # Smooth interpolation
    smooth_result = weight * log_barrier + (1 - weight) * combined_barrier

    return jnp.clip(smooth_result, 0, 1e8)
```

**Comment by lorenzo96-cmd** [comment]
Hi @LockedFlysher ,
This looks like a good improvement. Did you have the chance to test it?

**Comment by LockedFlysher** [comment]
Yes, it also works on H1 and talos example, seems the performance keeps the same as your code, a continuously differentiable function can be achieved.

**Comment by WangFu12138** [comment]
> Yes, it also works on H1 and talos example, seems the performance keeps the same as your code, a continuously differentiable function can be achieved.

Hi @LockedFlysher 
Thank you for maintaining such a fantastic project! I have a quick, perhaps somewhat naive question.

Is it currently possible to achieve full-pipeline automatic differentiation for an MPC + MuJoCo (via mujoco-mjx) setup using mpx in JAX? I'm trying to backpropagate gradients end-to-end through both the physics engine and the MPC solver for a differentiable control research project.

I would be extremely grateful for any brief advice or thoughts on whether mpx is naturally suited for this, or if it would be a very complex undertaking. Thank you for your time!

**Comment by lorenzo96-cmd** [comment]
Mpx is differentiable; you shouldn't have problems with it. On mjx side, be sure to use <option iterations="1" \> in the XML to set the number of Newton iterations to 1; otherwise, you will get an error due to the while loop. 
look at: 
https://github.com/google-deepmind/mujoco/issues/1182

**Comment by WangFu12138** [comment]
> Mpx is differentiable; you shouldn't have problems with it. On mjx side, be sure to use <option iterations="1" > in the XML to set the number of Newton iterations to 1; otherwise, you will get an error due to the while loop. look at: [google-deepmind/mujoco#1182](https://github.com/google-deepmind/mujoco/issues/1182)

Thank you for the quick and helpful response and for pointing me to that issue! I will definitely update the XML with <option iterations="1"> to bypass the while loop compilation error.

I do have a quick follow-up question regarding this workaround: won't limiting the solver to just a single Newton iteration significantly compromise the accuracy of contact resolution and constraint satisfaction? >
My concern is that this physics inaccuracy might negatively impact the execution performance and stability of the MPC. Are there common workarounds in mpx/mjx to mitigate this? For example, would you recommend drastically reducing the simulation timestep, or is there a way to force a fixed number of unrolled iterations (e.g., iterations="5") without triggering the while loop error?

Thanks again for your time and guidance!

**Comment by lorenzo96-cmd** [comment]
Reducing the number of iterations for sure reduces the accuracy of mujoco, but there are a few examples of controllers deployed using such an approximation, see:
- [dial mpc](https://github.com/LeCAR-Lab/dial-mpc) 
- [dynaFlow](https://arxiv.org/abs/2509.19804)
As far as setting the number of iterations to a fixed value, I guess that you will end up with the same problem; to avoid it you should change the while loop to a scan, but I don't suggest messing with mjx's implementation.
